pallets/pool-system/src/impls.rs

```diff
@@ -367,6 +372,60 @@ impl<T: Config> InvestmentAccountant<T::AccountId> for Pallet<T> {
 	}
 }
 
+impl<T: Config> ChangeGuard for Pallet<T> {
+	type Change = T::RuntimeChange;
+	type ChangeId = T::Hash;
+	type PoolId = T::PoolId;
+
+	fn note(pool_id: Self::PoolId, change: Self::Change) -> Result<Self::ChangeId, DispatchError> {
+		let registered_change = (Self::now(), change.clone());
+		let change_id: Self::ChangeId = T::Hashing::hash(&registered_change.encode());
+		Changes::<T>::insert(pool_id, change_id, registered_change);
+
+		Self::deposit_event(Event::ProposedChange {
+			pool_id,
+			change_id,
+			change,
+		});
+
+		Ok(change_id)
+	}
+
+	fn released(
+		pool_id: Self::PoolId,
+		change_id: Self::ChangeId,
+	) -> Result<Self::Change, DispatchError> {
+		let (submitted_time, change) =
+			Changes::<T>::get(pool_id, change_id).ok_or(Error::<T>::ChangeNotFound)?;
+
+		let pool_change: PoolChangeProposal = change.clone().into();
+
+		let mut allowed = true;
+		for requirement in pool_change.requirements() {
+			allowed &= match requirement {
+				Requirement::NextEpoch => {
+					match Pool::<T>::get(pool_id) {
+						Some(details) => submitted_time < details.epoch.last_closed,
+						None => true, // If the pool doesn't exists, there are no restrictions.
+					}
+				}
+				Requirement::DelayTime(secs) => {
+					Self::now().saturating_sub(submitted_time) >= secs as u64
+				}
+				// TODO
```

Comment:
Or better IMO. 

* pallet-investments stores some info about redemptions in the pallet-system. As opaque as the `RuntimeChange` for the pallet-pool-system. 
* But it allows to query whether there are redemtions outstainding

This would allow us to mitigate the need to call into pallet-investments and iterate over all tranches and possible outstanding redemptions. But just a first thought.

Reply:
Mmm... 🤔 Ok. 

> pallet-investments stores some info about redemptions in the pallet-system

How is it stored in the `pallet-system`? Is there a way to store things there?

Not sure if this is what you mean, but maybe `pallet-investments` can create a change (a `RuntimeChange`) as a way to send information to the `pallet-pool-system` in a transparent way. Although maybe it could be even simple to offer a trait from `pallet-pool-system` that `pallet-investment` can call to update the redemption state in the pool side

Reply:
I think it should check not if there are redemptions outstanding right now, but rather if all redemptions were fulfilled in the last epoch. Otherwise someone can block all changes indefinitely by continuously submitting a 1 wei redemption order in every block.

Reply:
Which was actually incorrectly implemented previously as well

Reply:
Ok, I need to dive first into this part that is new for me, but I get the idea, thanks 👍🏻

Reply:
> I think it should check not if there are redemptions outstanding right now, but rather if all redemptions were fulfilled in the last epoch.

But the change we are processing at that moment will affect the current redemptions, isn't it? So are the current redemptions which should be protected against the change.

I see the issue you commented on, but I'm not sure how it can fix it...

Reply:
> the last executed epoch had a redemption fulfillment of 100% in all tranches

But this information isn't anywhere right now, right? I guess we need to track this somehow to be able to check it later in the `released()` method.